src/binary_annotator.cpp

```diff
@@ -0,0 +1,647 @@
+
+  if (file_identifier_value) {
+    // Check if the file identifier region has non-zero data, and assume its the
+    // file identifier. Otherwise, it will get filled in with padding later.
+    regions.emplace_back(MakeBinaryRegion(
+        offset + sizeof(uint32_t), 4 * sizeof(uint8_t), BinaryRegionType::Char,
+        4, 0, std::string("File Identifier")));
+  }
+
+  sections_.insert(std::make_pair(
+      offset, MakeBinarySection("", BinarySectionType::Header, regions)));
+  return root_table_offset;
+}
+
+void BinaryAnnotator::BuildVTable(uint64_t offset,
+                                  const reflection::Object *table) {
+  const uint64_t vtable_offset = offset;
+
+  // First see if we have used this vtable before, if so skip building it again.
+  auto it = vtables_.find(vtable_offset);
+  if (it != vtables_.end()) { return; }
+
+  std::vector<BinaryRegion> regions;
+
+  // Vtables start with the size of the vtable
+  const uint16_t vtable_size = GetScalar<uint16_t>(offset);
+  regions.emplace_back(MakeBinaryRegion(offset, sizeof(uint16_t),
+                                        BinaryRegionType::Uint16, 0, 0,
+                                        std::string("size of this vtable")));
+  offset += sizeof(uint16_t);
+
+  // Then they have the size of the table they reference.
+  regions.emplace_back(
+      MakeBinaryRegion(offset, sizeof(uint16_t), BinaryRegionType::Uint16, 0, 0,
+                       std::string("size of referring table")));
+  offset += sizeof(uint16_t);
+
+  const uint64_t offset_start = offset;
+
+  // A mapping between field (and its id) to the relative offset (uin16_t) from
+  // the start of the table.
+  std::vector<std::tuple<const reflection::Field *, uint16_t>> fields;
+
+  // Loop over all the fields.
+  ForAllFields(table, /*reverse=*/false, [&](const reflection::Field *field) {
+    const uint64_t field_offset = offset_start + field->id() * sizeof(uint16_t);
+
+    if (field_offset >= vtable_offset + vtable_size) {
+      // This field_offset is too large for this vtable, so it must come from a
+      // newer schema than the binary was create with. Ignore.
+
+      // TODO(dbaileychess): We could show which fields are not set an their
+      // default values if we want. We just need a way to make it obvious that
+      // it isn't part of the buffer.
+      return;
+    }
+
+    const uint16_t value = GetScalar<uint16_t>(field_offset);
+
+    fields.push_back(std::make_pair(field, GetScalar<uint16_t>(field_offset)));
+    regions.push_back(MakeBinaryRegion(
+        field_offset, sizeof(uint16_t), BinaryRegionType::VOffset, 0, 0,
+        std::string("offset to field `") + field->name()->c_str() +
+            "` (id: " + std::to_string(field->id()) + ")" +
+            ((value == 0) ? " <defaults to " +
+                                std::to_string(field->default_integer()) + ">"
+                          : "")));
+  });
+
+  sections_.insert(std::make_pair(
```

Comment:
I think you don't need `insert` and can use `operator[]`  since we check for a new vtable at the start.

Reply:
Also consider working "in place" so you don't have to insert at all

e.g.
```cpp
std::vector<...>& fields = vtables[vtable_offset].fields;
```

Reply:
Done.

I move the fields, so there is no real penalty to do it as is.